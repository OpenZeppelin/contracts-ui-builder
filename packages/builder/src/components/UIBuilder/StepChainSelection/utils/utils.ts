import type { NetworkConfig } from '@openzeppelin/contracts-ui-builder-types';

export const ICON_SIZE = 16;

export function getNetworkIconName(network: NetworkConfig): string | null {
  if (network.ecosystem === 'midnight') {
    return null;
  }
  return network.icon || network.network.toLowerCase();
}
