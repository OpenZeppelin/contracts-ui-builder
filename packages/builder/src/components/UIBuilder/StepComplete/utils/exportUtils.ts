/**
 * Creates a downloadable file from a Blob or ArrayBuffer.
 * @param blob - The blob data to download
 * @param fileName - The name to give the downloaded file
 */
export function downloadBlob(blob: Blob, fileName: string): void {
  const url = window.URL.createObjectURL(blob);
  const a = document.createElement('a');
  a.href = url;
  a.download = fileName;
  document.body.appendChild(a);
  a.click();
  setTimeout(() => {
    document.body.removeChild(a);
    window.URL.revokeObjectURL(url);
  }, 0);
}
