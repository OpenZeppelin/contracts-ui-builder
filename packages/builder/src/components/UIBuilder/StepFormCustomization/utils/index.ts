export * from './fieldEditorUtils';
export * from './fieldTypeUtils';
export * from './executionUtils';
