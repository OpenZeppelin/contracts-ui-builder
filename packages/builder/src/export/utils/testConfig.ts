import { capitalize } from 'lodash';

import { Ecosystem } from '@openzeppelin/ui-builder-types';
import type { ContractSchema, FieldType } from '@openzeppelin/ui-builder-types';
import { generateId } from '@openzeppelin/ui-builder-utils';

import type { BuilderFormConfig } from '../../core/types/FormTypes';

/**
 * Create a minimal form configuration for testing
 * @param functionName - Optional function name (defaults to 'testFunction')
 * @param ecosystem - Optional ecosystem (defaults to 'evm')
 * @returns A minimal BuilderFormConfig instance for testing
 */
export function createMinimalFormConfig(
  functionName: string = 'testFunction',
  _ecosystem: Ecosystem = 'evm'
): BuilderFormConfig {
  return {
    functionId: functionName,
    fields: [
      {
        id: generateId(),
        type: 'text',
        name: 'testParam',
        label: 'Test Parameter',
        validation: {
          required: true,
        },
        helperText: `Description for Test Parameter`,
        placeholder: 'Enter test parameter',
      },
    ],
    layout: {
      columns: 1,
      spacing: 'normal',
      labelPosition: 'top',
    },
    validation: {
      mode: 'onChange',
      showErrors: 'inline',
    },
    executionConfig: {
      method: 'eoa',
      allowAny: true,
    },
    theme: {},
    contractAddress: '0xe34139463bA50bD61336E0c446Bd8C0867c6fE65',
  };
}

/**
 * Creates a minimal mock ContractSchema for testing.
 */
export function createMinimalContractSchema(
  functionName: string,
  ecosystem: Ecosystem
): ContractSchema {
  return {
    ecosystem,
    name: 'MockContract',
    address: '0x1234567890123456789012345678901234567890', // Use a valid-looking address
    functions: [
      {
        id: functionName,
        name: functionName,
        displayName: capitalize(functionName),
        inputs: [], // Basic schema assumes no inputs for simplicity
        type: 'function',
        modifiesState: true,
      },
      // Add a basic view function for potential widget tests?
      {
        id: 'viewFunction',
        name: 'viewFunction',
        displayName: 'View Function',
        inputs: [],
        type: 'function',
        modifiesState: false,
        stateMutability: 'view',
      },
    ],
  };
}

/**
 * Create a complex form configuration for testing
 * @param functionName - Optional function name (defaults to 'complexFunction')
 * @param ecosystem - Optional ecosystem (defaults to 'evm')
 * @returns A complex BuilderFormConfig instance with multiple field types
 */
export function createComplexFormConfig(
  functionName: string = 'complexFunction',
  _ecosystem: Ecosystem = 'evm'
): BuilderFormConfig {
  const basicFieldIds = [generateId(), generateId(), generateId()];
  const advancedFieldIds = [generateId(), generateId(), generateId(), generateId()];

  return {
    functionId: functionName,
    title: 'Complex Test Form',
    description: 'A test form with multiple different field types',
    fields: [
      {
        id: basicFieldIds[0],
        type: 'text',
        name: 'stringParam',
        label: 'Text Parameter',
        placeholder: 'Enter text',
        helperText: 'A basic text field',
        validation: {
          required: true,
          minLength: 3,
          maxLength: 100,
        },
      },
      {
        id: basicFieldIds[1],
        type: 'number',
        name: 'numberParam',
        label: 'Number Parameter',
        placeholder: 'Enter a number',
        helperText: 'A number field',
        validation: {
          required: true,
          min: 1,
          max: 100,
        },
      },
      {
        id: basicFieldIds[2],
        type: 'checkbox',
        name: 'boolParam',
        label: 'Boolean Parameter',
        helperText: 'A checkbox field',
        defaultValue: false,
        validation: {
          required: true,
        },
      },
      {
        id: advancedFieldIds[0],
        type: 'blockchain-address',
        name: 'addressParam',
        label: 'Address Parameter',
        placeholder: '0x...',
        helperText: 'Enter an Ethereum address',
        validation: {
          required: true,
        },
      },
      {
        id: advancedFieldIds[1],
        type: 'textarea',
        name: 'longTextParam',
        label: 'Long Text Parameter',
        placeholder: 'Enter detailed information',
        helperText: 'A textarea for longer content',
        validation: {
          required: false,
          maxLength: 1000,
        },
      },
      {
        id: advancedFieldIds[2],
        type: 'select',
        name: 'selectParam',
        label: 'Select Parameter',
        placeholder: 'Choose an option',
        helperText: 'A dropdown selection field',
        validation: {
          required: true,
        },
        options: [
          { label: 'Option A', value: 'a' },
          { label: 'Option B', value: 'b' },
          { label: 'Option C', value: 'c' },
        ],
      },
      {
        id: advancedFieldIds[3],
        type: 'date',
        name: 'dateParam',
        label: 'Date Parameter',
        placeholder: 'Select a date',
        helperText: 'A date selection field',
        validation: {
          required: false,
        },
      },
      {
        id: 'hardcodedTextParam',
        type: 'text',
        name: 'hardcodedTextParam',
        label: 'Hardcoded Text',
        isHardcoded: true,
        hardcodedValue: 'This is a fixed value',
        helperText: 'This field has a hardcoded value',
        validation: {},
      },
      {
        id: 'readOnlyText',
        type: 'text',
        name: 'readOnlyText',
        label: 'Read-Only Text',
        readOnly: true,
        defaultValue: 'Cannot be changed',
        helperText: 'This field is read-only',
        validation: {},
      },
      {
        id: 'readOnlyHardcodedNumber',
        type: 'number',
        name: 'readOnlyHardcodedNumber',
        label: 'Hardcoded Read-Only Number',
        isHardcoded: true,
        hardcodedValue: 42,
        readOnly: true,
        helperText: 'This field is hardcoded and read-only',
        validation: {},
      },
    ],
    layout: {
      columns: 1,
      spacing: 'normal',
      labelPosition: 'top',
    },
    validation: {
      mode: 'onChange',
      showErrors: 'inline',
    },
    executionConfig: {
      method: 'eoa',
      allowAny: false,
      specificAddress: '0x0000000000000000000000000000000000000000',
    },
    theme: {},
    contractAddress: '0xe34139463bA50bD61336E0c446Bd8C0867c6fE65',
  };
}

/**
 * Create a test field
 * @param type - The field type
 * @param name - The field name
 * @param label - The field label
 * @param options - Optional field options
 * @returns A field configuration object
 */
export function createTestField(
  type: FieldType,
  name: string,
  label: string,
  options: Record<string, unknown> = {}
): {
  id: string;
  type: FieldType;
  name: string;
  label: string;
  validation: { required: boolean };
  helperText: string;
  placeholder: string;
  [key: string]: unknown;
} {
  return {
    id: generateId(),
    type,
    name,
    label,
    validation: {
      required: true,
    },
    helperText: `Description for ${label}`,
    placeholder: `Enter ${label.toLowerCase()}`,
    ...options,
    contractAddress: '0xe34139463bA50bD61336E0c446Bd8C0867c6fE65',
  };
}
