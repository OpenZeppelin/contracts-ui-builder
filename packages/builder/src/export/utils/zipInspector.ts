import JSZip from 'jszip';

import { logger } from '@openzeppelin/ui-builder-utils';

/**
 * Optional callback for testing error handling during file extraction.
 * @param filePath The path of the file being processed.
 * @param fileObject The JSZip file object.
 * @returns void | Promise<void> - If it throws an error, extraction for that file will fail.
 */
export type ZipExtractionTestCallback = (
  filePath: string,
  fileObject: JSZip.JSZipObject
) => void | Promise<void>;

/**
 * Extracts all files from a ZIP blob into a record of paths and string content.
 * Handles potential extraction errors gracefully.
 * @param zipBlob The ZIP file content as a Blob.
 * @param testCallback Optional callback for simulating errors during testing.
 */
export async function extractFilesFromZip(
  zipData: Blob | Buffer,
  testCallback?: ZipExtractionTestCallback
): Promise<Record<string, string>> {
  const files: Record<string, string> = {};
  try {
    const zip = await JSZip.loadAsync(zipData);

    // Create an array of promises for file extraction
    const extractionPromises = Object.entries(zip.files).map(async ([path, fileObject]) => {
      // Skip directories
      if (fileObject.dir) {
        return;
      }

      try {
        // --- TEST HOOK START ---
        // If a test callback is provided, execute it here.
        // If it throws, the catch block below will handle it.
        if (testCallback) {
          await testCallback(path, fileObject);
        }
        // --- TEST HOOK END ---

        // Extract file content as text
        files[path] = await fileObject.async('text');
      } catch (error) {
        // Log the error and store an error message in the files record
        const errorMessage = `ERROR_EXTRACTING: ${(error as Error).message}`;
        logger.error('zipInspector', `Error extracting file ${path}:`, error); // Use logger
        files[path] = errorMessage;
      }
    });

    // Wait for all extraction attempts to complete
    await Promise.all(extractionPromises);
  } catch (error) {
    // Log error if the ZIP itself cannot be loaded
    logger.error('zipInspector', 'Error loading or processing ZIP file:', error);
    // You might want to throw this error or handle it differently
    // For now, return the potentially partially populated files object or an empty one
  }
  return files;
}

/**
 * Options for validating an exported project
 */
export interface ProjectValidationOptions {
  /** List of files that must exist in the export */
  requiredFiles?: string[];
  /** List of files that should not exist in the export */
  forbiddenFiles?: string[];
  /** Object mapping file paths to validation functions */
  contentValidations?: Record<string, (content: string) => boolean | string>;
  /** Object mapping file paths to regex patterns that must match */
  contentPatterns?: Record<string, RegExp>;
}

/**
 * Result of validating an exported project
 */
export interface ProjectValidationResult {
  /** Whether the validation passed */
  isValid: boolean;
  /** List of validation errors */
  errors: string[];
  /** Information about the validation that passed */
  successes: string[];
}

/**
 * Validate an exported project against expectations
 * @param files - The files extracted from the ZIP
 * @param options - Validation options
 * @returns Validation result
 */
export function validateExportedProject(
  files: Record<string, string>,
  options: ProjectValidationOptions
): ProjectValidationResult {
  const errors: string[] = [];
  const successes: string[] = [];

  // Check required files
  if (options.requiredFiles) {
    for (const file of options.requiredFiles) {
      if (files[file] === undefined) {
        errors.push(`Missing required file: ${file}`);
      } else {
        successes.push(`Found required file: ${file}`);
      }
    }
  }

  // Check forbidden files
  if (options.forbiddenFiles) {
    for (const file of options.forbiddenFiles) {
      if (files[file] !== undefined) {
        errors.push(`Found forbidden file: ${file}`);
      } else {
        successes.push(`Correctly missing forbidden file: ${file}`);
      }
    }
  }

  // Check content validations
  if (options.contentValidations) {
    for (const [file, validator] of Object.entries(options.contentValidations)) {
      if (files[file] !== undefined) {
        const result = validator(files[file]);
        if (result !== true) {
          const message =
            typeof result === 'string' ? result : `Content validation failed for ${file}`;
          errors.push(message);
        } else {
          successes.push(`Content validation passed for ${file}`);
        }
      } else {
        errors.push(`Cannot validate missing file: ${file}`);
      }
    }
  }

  // Check regex patterns
  if (options.contentPatterns) {
    for (const [file, pattern] of Object.entries(options.contentPatterns)) {
      if (files[file] !== undefined) {
        if (!pattern.test(files[file])) {
          errors.push(`Pattern match failed for ${file}: ${pattern}`);
        } else {
          successes.push(`Pattern match passed for ${file}`);
        }
      } else {
        errors.push(`Cannot validate pattern on missing file: ${file}`);
      }
    }
  }

  return {
    isValid: errors.length === 0,
    errors,
    successes,
  };
}
