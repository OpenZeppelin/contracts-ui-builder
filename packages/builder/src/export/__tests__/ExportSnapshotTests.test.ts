import { afterEach, beforeEach, describe, expect, it } from 'vitest';

import { Ecosystem, EvmNetworkConfig } from '@openzeppelin/ui-builder-types';
import { logger } from '@openzeppelin/ui-builder-utils';

import { AppExportSystem } from '../AppExportSystem';
import { createMinimalContractSchema, createMinimalFormConfig } from '../utils/testConfig';
import { extractFilesFromZip } from '../utils/zipInspector';

// Define mock network config
const mockEvmNetworkConfig: EvmNetworkConfig = {
  id: 'test-export-snapshot-evm',
  name: 'Test Export Snapshot EVM',
  exportConstName: 'mockEvmNetworkConfig',
  ecosystem: 'evm',
  network: 'ethereum',
  type: 'testnet',
  isTestnet: true,
  chainId: 1337,
  rpcUrl: 'http://localhost:8545',
  nativeCurrency: { name: 'TETH', symbol: 'TETH', decimals: 18 },
  apiUrl: '',
};

describe('Export Snapshot Tests', () => {
  /**
   * Helper function to extract key files from the export for snapshot testing
   */
  async function getSnapshotFiles(ecosystem: Ecosystem = 'evm', functionName: string = 'transfer') {
    // Create the export system and form config
    const exportSystem = new AppExportSystem();
    const formConfig = createMinimalFormConfig(functionName, ecosystem);
    const mockContractSchema = createMinimalContractSchema(functionName, ecosystem);

    // Export the form with a consistent project name for snapshots
    const result = await exportSystem.exportApp(
      formConfig,
      mockContractSchema,
      mockEvmNetworkConfig,
      functionName,
      {
        projectName: 'snapshot-test-project',
      }
    );

    // Extract files from the ZIP using result.data
    expect(result.data).toBeDefined();
    const files = await extractFilesFromZip(result.data);

    // Return the files that should be snapshot tested
    return {
      packageJson: JSON.parse(files['package.json']),
      appComponent: files['src/App.tsx'],
      formComponent: files['src/components/GeneratedForm.tsx'],
      adapterIndex: files['src/adapters/index.ts'],
      [`adapter_${ecosystem}`]: files[`src/adapters/${ecosystem}/adapter.ts`],
    };
  }

  /**
   * Filter sensitive or variable content from snapshots to avoid false diffs
   */
  function prepareForSnapshot(content: string): string {
    if (!content) return '';

    // Replace timestamp comments that would cause snapshot diffs
    return (
      content
        .replace(/\/\/ Generated at:.*$/gm, '// Generated at: [timestamp]')
        .replace(/\/\* Generated:.*\*\//gm, '/* Generated: [timestamp] */')
        // Replace IDs in the form schema - more comprehensive UUID pattern
        .replace(
          /"id":\s*"[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}"/g,
          '"id": "[id]"'
        )
        // Replace field IDs directly in content - better UUID pattern matching
        .replace(
          /id: '[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}'/g,
          "id: '[id]'"
        )
        // Replace fields list in sections
        .replace(/"fields":\s*\[\s*"[a-zA-Z0-9-]+"\s*\]/g, '"fields": ["testParam"]')
        // Replace other variables that might make snapshots unstable
        .replace(/version: ["'][\d.]+["']/g, 'version: "[version]"')
    );
  }

  describe('EVM Export Snapshots', () => {
    it('should match snapshot for App component', async () => {
      const files = await getSnapshotFiles('evm');
      expect(prepareForSnapshot(files.appComponent)).toMatchSnapshot('app-component-evm');
    });

    it('should match snapshot for Form component', async () => {
      const files = await getSnapshotFiles('evm');
      expect(prepareForSnapshot(files.formComponent)).toMatchSnapshot('form-component-evm');
    });

    it('should match snapshot for adapters index', async () => {
      const files = await getSnapshotFiles('evm');
      expect(prepareForSnapshot(files.adapterIndex)).toMatchSnapshot('adapter-index-evm');
    });

    it('should match snapshot for EVM adapter', async () => {
      const files = await getSnapshotFiles('evm');
      expect(prepareForSnapshot(files.adapter_evm)).toMatchSnapshot('evm-adapter');
    });

    it('should match snapshot for package.json structure', async () => {
      const files = await getSnapshotFiles('evm');

      // Only snapshot relevant parts of package.json to avoid noise
      const { dependencies, devDependencies, scripts } = files.packageJson;
      expect({ dependencies, devDependencies, scripts }).toMatchSnapshot('package-json-evm');
    });
  });

  describe('Solana Export Snapshots', () => {
    it('should match snapshot for Solana adapter', async () => {
      const files = await getSnapshotFiles('solana');
      expect(prepareForSnapshot(files.adapter_solana)).toMatchSnapshot('solana-adapter');
    });

    it('should match snapshot for adapters index with Solana', async () => {
      const files = await getSnapshotFiles('solana');
      expect(prepareForSnapshot(files.adapterIndex)).toMatchSnapshot('adapter-index-solana');
    });

    it('should match snapshot for package.json with Solana dependencies', async () => {
      const files = await getSnapshotFiles('solana');

      // Only snapshot relevant parts of package.json to avoid noise
      const { dependencies, devDependencies, scripts } = files.packageJson;
      expect({ dependencies, devDependencies, scripts }).toMatchSnapshot('package-json-solana');
    });
  });

  describe('Cross-Chain Comparison', () => {
    it('should verify form component similarities across chains', async () => {
      // Get form components for different chains
      const evmFiles = await getSnapshotFiles('evm');
      const solanaFiles = await getSnapshotFiles('solana');

      // The core form structure should be similar
      // Looking at component definition and main return elements
      const evmFormComponent = prepareForSnapshot(evmFiles.formComponent);
      const solanaFormComponent = prepareForSnapshot(solanaFiles.formComponent);

      // Both should be React components
      expect(evmFormComponent).toContain('import');
      expect(evmFormComponent).toContain("from 'react'");
      expect(solanaFormComponent).toContain('import');
      expect(solanaFormComponent).toContain("from 'react'");

      // Both should render TransactionForm
      expect(evmFormComponent).toMatch(/return\s*\([\s\S]*?<TransactionForm/);
      expect(solanaFormComponent).toMatch(/return\s*\([\s\S]*?<TransactionForm/);

      // Both should have similar form schema structures
      expect(evmFormComponent).toMatch(/const formSchema/);
      expect(solanaFormComponent).toMatch(/const formSchema/);
    });

    it('should verify App component similarities across chains', async () => {
      // Get App components for different chains
      const evmFiles = await getSnapshotFiles('evm');
      const solanaFiles = await getSnapshotFiles('solana');

      // The App component should be nearly identical across chains
      const evmAppComponent = prepareForSnapshot(evmFiles.appComponent);
      const solanaAppComponent = prepareForSnapshot(solanaFiles.appComponent);

      // Both should import and render GeneratedForm
      expect(evmAppComponent).toMatch(/import.*GeneratedForm/);
      expect(solanaAppComponent).toMatch(/import.*GeneratedForm/);

      // Both should have similar App component structure
      expect(evmAppComponent).toMatch(/function App/);
      expect(solanaAppComponent).toMatch(/function App/);
    });
  });
});

// --- Add New Test Suite for Conditional Files ---

describe('Export Snapshot Tests > Conditional File Modifications', () => {
  beforeEach(() => {
    // Disable logging for cleaner test output
    logger.configure({ enabled: false });
  });

  afterEach(() => {
    // Restore default logger settings
    logger.configure({ enabled: true, level: 'info' });
  });

  it('should modify styles.css import correctly based on isCliBuildTarget option', async () => {
    const mainCssPath = 'src/styles.css';
    const startingCssContent = `@import 'tailwindcss';\n@import './styles/global.css';\n\n/* Base styles... */`;

    // --- Simulate Case 1: isCliBuildTarget = true ---
    let projectFilesCli: Record<string, string> = { [mainCssPath]: startingCssContent };
    // Directly apply the modification logic here
    const originalCliContent = projectFilesCli[mainCssPath];
    const modifiedCliContent = originalCliContent.replace(
      /^\s*@import\s+['"]tailwindcss['"]\s*;?/m,
      "@import 'tailwindcss' source('../../../');"
    );
    if (modifiedCliContent !== originalCliContent) {
      projectFilesCli[mainCssPath] = modifiedCliContent;
    } else {
      // Fail test if replacement didn't happen, indicating regex issue
      throw new Error('CSS replacement failed for CLI case');
    }
    const stylesCssCli = projectFilesCli[mainCssPath];
    expect(stylesCssCli).toBeDefined();
    expect(stylesCssCli).toMatchSnapshot('styles-css-cli');

    // --- Simulate Case 2: isCliBuildTarget = false ---
    let projectFilesUi: Record<string, string> = { [mainCssPath]: startingCssContent };
    // In the UI case, we expect NO replacement, so modified should equal original
    // We check this implicitly by snapshotting the *original* content effectively
    const stylesCssUi = projectFilesUi[mainCssPath]; // Snapshot the unmodified content
    expect(stylesCssUi).toBeDefined();
    expect(stylesCssUi).toMatchSnapshot('styles-css-ui');

    // --- Final Assertions ---
    // Verify the content is different
    expect(stylesCssCli).not.toEqual(stylesCssUi);
    // Verify CLI version contains the source directive
    expect(stylesCssCli).toContain("source('../../../')");
    // Verify UI version does NOT contain the source directive
    expect(stylesCssUi).not.toContain("source('../../../')");
    expect(stylesCssUi).toEqual(startingCssContent); // Ensure UI case is unchanged
  });
});
