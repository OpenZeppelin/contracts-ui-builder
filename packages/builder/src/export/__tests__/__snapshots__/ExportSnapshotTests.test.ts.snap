// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Export Snapshot Tests > Conditional File Modifications > should modify styles.css import correctly based on isCliBuildTarget option > styles-css-cli 1`] = `
"@import 'tailwindcss' source('../../../');
@import './styles/global.css';

/* Base styles... */"
`;

exports[`Export Snapshot Tests > Conditional File Modifications > should modify styles.css import correctly based on isCliBuildTarget option > styles-css-ui 1`] = `
"@import 'tailwindcss';
@import './styles/global.css';

/* Base styles... */"
`;

exports[`Export Snapshot Tests > EVM Export Snapshots > should match snapshot for App component > app-component-evm 1`] = `
"import {
  useDerivedAccountStatus,
  useWalletState,
  WalletConnectionWithSettings,
} from '@openzeppelin/contracts-ui-builder-react-core';
import { Footer } from '@openzeppelin/contracts-ui-builder-ui';

import GeneratedForm from './components/GeneratedForm';

/**
 * App Component
 *
 * Main application component that wraps the form.
 * Uses useWalletState to get the active adapter.
 */
export function App() {
  const { activeAdapter, isAdapterLoading } = useWalletState();
  const { isConnected: isWalletConnectedForForm } = useDerivedAccountStatus();

  if (isAdapterLoading) {
    return <div className="app-loading">Loading adapter...</div>;
  }
  if (!activeAdapter) {
    return (
      <div className="app-error">
        Adapter not available. Please ensure network is selected and supported.
      </div>
    );
  }

  return (
    <div className="app">
      <header className="header border-b px-6 py-3">
        <div className="container mx-auto flex items-center justify-between">
          <div className="flex items-center gap-4">
            <img src="/OZ-Logo-BlackBG.svg" alt="OpenZeppelin Logo" className="h-6 w-auto" />
            <div className="h-5 border-l border-gray-300 mx-1"></div>
            <div>
              <h1 className="text-base font-medium">transfer</h1>
              <p className="text-xs text-muted-foreground">
                Form for interacting with blockchain contracts
              </p>
            </div>
          </div>
          <WalletConnectionWithSettings />
        </div>
      </header>

      <main className="main">
        <div className="container">
          <GeneratedForm adapter={activeAdapter} isWalletConnected={isWalletConnectedForForm} />
        </div>
      </main>

      <Footer />
    </div>
  );
}
"
`;

exports[`Export Snapshot Tests > EVM Export Snapshots > should match snapshot for EVM adapter > evm-adapter 1`] = `""`;

exports[`Export Snapshot Tests > EVM Export Snapshots > should match snapshot for Form component > form-component-evm 1`] = `
"import { useState } from 'react';

import {
  ContractActionBar,
  ContractStateWidget,
  TransactionForm,
} from '@openzeppelin/contracts-ui-builder-renderer';
import type {
  ContractAdapter,
  ContractSchema,
  ExecutionConfig,
  RenderFormSchema,
} from '@openzeppelin/contracts-ui-builder-types';
import { Card, CardContent } from '@openzeppelin/contracts-ui-builder-ui';

// Props for GeneratedForm
interface GeneratedFormProps {
  adapter: ContractAdapter;
  isWalletConnected?: boolean;
}

/**
 * Generated Transaction Form for transfer
 *
 * This component renders a form for interacting with a blockchain contract.
 * It uses the shared renderer package which ensures consistent behavior
 * with the preview in the builder app.
 */
export default function GeneratedForm({ adapter, isWalletConnected }: GeneratedFormProps) {
  // TODO: Enable this useEffect as a fallback?
  // If the adapter supports runtime schema loading (e.g., via Etherscan)
  // and the injected schema is missing or invalid, this could attempt to load it.
  /*
  const [contractSchema, setContractSchema] = useState<ContractSchema | null>(null);
  */
  const [isWidgetVisible, setIsWidgetVisible] = useState(false);
  const [loadError, _setLoadError] = useState<Error | null>(null);
  // Form schema generated from the builder and transformed by FormSchemaFactory
  const formSchema: RenderFormSchema = {
    functionId: 'transfer',
    layout: {
      columns: 1,
      spacing: 'normal',
      labelPosition: 'top',
    },
    validation: {
      mode: 'onChange',
      showErrors: 'inline',
    },
    theme: {},
    contractAddress: '0xe34139463bA50bD61336E0c446Bd8C0867c6fE65',
    id: 'form-transfer',
    title: 'Transfer',
    description: 'Form for interacting with the Transfer function.',
    fields: [
      {
        id: '[id]',
        name: 'testParam',
        label: 'Test Parameter',
        type: 'text',
        placeholder: 'Enter test parameter',
        helperText: 'Description for Test Parameter',
        validation: {
          required: true,
        },
      },
    ],
    submitButton: {
      text: 'Execute Transfer',
      loadingText: 'Processing...',
      variant: 'primary',
    },
  }; /*@@FORM_SCHEMA_JSON@@*/

  // Contract schema injected by generator (loaded or imported by the user)
  const contractSchema: ContractSchema = {
    ecosystem: 'evm',
    name: 'MockContract',
    address: '0x1234567890123456789012345678901234567890',
    functions: [
      {
        id: 'transfer',
        name: 'transfer',
        displayName: 'Transfer',
        inputs: [],
        type: 'function',
        modifiesState: true,
      },
      {
        id: 'viewFunction',
        name: 'viewFunction',
        displayName: 'View Function',
        inputs: [],
        type: 'function',
        modifiesState: false,
        stateMutability: 'view',
      },
    ],
  };

  // Execution configuration selected in the builder
  const executionConfig: ExecutionConfig | undefined = {
    method: 'eoa',
    allowAny: true,
  };

  const contractAddress = formSchema.contractAddress;

  // TODO: Enable this useEffect as a fallback?
  // If the adapter supports runtime schema loading (e.g., via Etherscan)
  // and the injected schema is missing or invalid, this could attempt to load it.
  /*
  useEffect(() => {
    setLoadError(null);
    setContractSchema(null);

    if (contractAddress) {
      adapter
        .loadContract(contractAddress)
        .then(setContractSchema)
        .catch((err: unknown) => {
          // Catch error during contract loading
          logger.error('GeneratedForm', 'Error loading contract schema:', err);
          // Create a new Error object if caught value is not already one
          const errorToSet =
            err instanceof Error ? err : new Error('Failed to load contract state');
          setLoadError(errorToSet);
          setContractSchema(null);
        });
    } else {
      setContractSchema(null);
    }
  }, [contractAddress, adapter]);
  */

  // Decide which schema to use: prioritize injected, fallback maybe later?
  const schemaToUse = contractSchema; // Sticking to injected schema for now

  const toggleWidget = () => {
    setIsWidgetVisible((prev: boolean) => !prev);
  };

  return (
    <div className="space-y-6">
      {/* Contract Action Bar - consistent with builder app */}
      {adapter.networkConfig && (
        <ContractActionBar
          networkConfig={adapter.networkConfig}
          contractAddress={contractAddress}
          onToggleContractState={toggleWidget}
          isWidgetExpanded={isWidgetVisible}
        />
      )}

      <div className="flex gap-4">
        {/* Contract State Widget on the left side - matching builder app layout */}
        {contractAddress && isWidgetVisible && (
          <div className="w-[300px] flex-shrink-0">
            <div className="sticky top-4">
              <ContractStateWidget
                contractSchema={schemaToUse}
                contractAddress={contractAddress}
                adapter={adapter}
                isVisible={isWidgetVisible}
                onToggle={toggleWidget}
                error={loadError}
              />
            </div>
          </div>
        )}

        {/* Main form on the right side */}
        <div className="flex-1">
          <Card>
            <CardContent className="space-y-4">
              <TransactionForm
                schema={formSchema}
                contractSchema={contractSchema}
                adapter={adapter}
                isWalletConnected={isWalletConnected}
                executionConfig={executionConfig}
              />
            </CardContent>
          </Card>
        </div>
      </div>
    </div>
  );
}
"
`;

exports[`Export Snapshot Tests > EVM Export Snapshots > should match snapshot for adapters index > adapter-index-evm 1`] = `""`;

exports[`Export Snapshot Tests > EVM Export Snapshots > should match snapshot for package.json structure > package-json-evm 1`] = `
{
  "dependencies": {
    "@openzeppelin/contracts-ui-builder-adapter-evm": "^0.8.0",
    "@openzeppelin/contracts-ui-builder-react-core": "^0.8.0",
    "@openzeppelin/contracts-ui-builder-renderer": "^0.8.0",
    "@openzeppelin/contracts-ui-builder-types": "^0.8.0",
    "@openzeppelin/contracts-ui-builder-ui": "^0.8.0",
    "@openzeppelin/contracts-ui-builder-utils": "^0.8.0",
    "@tanstack/react-query": "^5.0.0",
    "@wagmi/core": "^2.20.3",
    "react": "^19.0.0",
    "react-dom": "^19.0.0",
    "react-hook-form": "^7.60.0",
    "viem": "^2.28.0",
    "wagmi": "^2.15.0",
  },
  "devDependencies": {
    "@tailwindcss/vite": "^4.1.11",
    "@types/lodash": "^4.17.5",
    "@types/react": "^19.0.10",
    "@types/react-dom": "^19.0.4",
    "@typescript-eslint/eslint-plugin": "^8.26.0",
    "@typescript-eslint/parser": "^8.26.0",
    "@vitejs/plugin-react": "^4.3.1",
    "eslint": "^9.3.0",
    "eslint-plugin-react-hooks": "^5.0.0",
    "eslint-plugin-react-refresh": "^0.4.5",
    "tailwindcss": "^4.0.0",
    "typescript": "^5.8.2",
    "vite": "^6.2.5",
  },
  "scripts": {
    "build": "tsc && vite build",
    "check-deps": "npm outdated",
    "dev": "vite",
    "lint": "eslint . --ext ts,tsx --report-unused-disable-directives --max-warnings 0",
    "preview": "vite preview",
    "update-renderer": "npm update @openzeppelin/contracts-ui-builder-renderer",
  },
}
`;

exports[`Export Snapshot Tests > Solana Export Snapshots > should match snapshot for Solana adapter > solana-adapter 1`] = `""`;

exports[`Export Snapshot Tests > Solana Export Snapshots > should match snapshot for adapters index with Solana > adapter-index-solana 1`] = `""`;

exports[`Export Snapshot Tests > Solana Export Snapshots > should match snapshot for package.json with Solana dependencies > package-json-solana 1`] = `
{
  "dependencies": {
    "@openzeppelin/contracts-ui-builder-adapter-evm": "^0.8.0",
    "@openzeppelin/contracts-ui-builder-react-core": "^0.8.0",
    "@openzeppelin/contracts-ui-builder-renderer": "^0.8.0",
    "@openzeppelin/contracts-ui-builder-types": "^0.8.0",
    "@openzeppelin/contracts-ui-builder-ui": "^0.8.0",
    "@openzeppelin/contracts-ui-builder-utils": "^0.8.0",
    "@tanstack/react-query": "^5.0.0",
    "@wagmi/core": "^2.20.3",
    "react": "^19.0.0",
    "react-dom": "^19.0.0",
    "react-hook-form": "^7.60.0",
    "viem": "^2.28.0",
    "wagmi": "^2.15.0",
  },
  "devDependencies": {
    "@tailwindcss/vite": "^4.1.11",
    "@types/lodash": "^4.17.5",
    "@types/react": "^19.0.10",
    "@types/react-dom": "^19.0.4",
    "@typescript-eslint/eslint-plugin": "^8.26.0",
    "@typescript-eslint/parser": "^8.26.0",
    "@vitejs/plugin-react": "^4.3.1",
    "eslint": "^9.3.0",
    "eslint-plugin-react-hooks": "^5.0.0",
    "eslint-plugin-react-refresh": "^0.4.5",
    "tailwindcss": "^4.0.0",
    "typescript": "^5.8.2",
    "vite": "^6.2.5",
  },
  "scripts": {
    "build": "tsc && vite build",
    "check-deps": "npm outdated",
    "dev": "vite",
    "lint": "eslint . --ext ts,tsx --report-unused-disable-directives --max-warnings 0",
    "preview": "vite preview",
    "update-renderer": "npm update @openzeppelin/contracts-ui-builder-renderer",
  },
}
`;
